text/3503-frontmatter.md

```diff
@@ -0,0 +1,512 @@
+Like with Python, this allows any file in a package (with the correct deps and `mod`s) to be executed, allowing easier interacting experiences in verifying behavior.
+
+As for the hard-coded infostring used by cargo, that is a decision for [RFC 3502](https://github.com/rust-lang/rfcs/pull/3502).
+
+## Syntax
+
+When choosing a syntax for embedding manifests in source files, our care-abouts are
+- How obvious it is for new users when they see it
+- How easy it is for newer users to remember it and type it out
+- How machine editable it is for `cargo add` and friends
+- Needs to be valid Rust code for quality of error messages, etc
+- Leave the door open in case we want to reuse the syntax for embedded lockfiles
+- Leave the door open for single-file `lib`s
+
+This proposed syntax builds off of the precedence of Rust syntax being specialized for an external tool
+(doc-comments for rustdoc).
+However, a new syntax is used instead of extending the comment syntax:
+- Focused on being parsed by arbitrary tools (cargo, vim plugins, etc) without understanding the full Rust grammar
+- Side steps compatibility issues with both
+  user expectations with the looseness of comment syntax (making it harder to parse)
+  and any existing comments that may look like a new structured comment syntax
+
+Benefits for the proposed solution are:
+- Visually/syntactically lightweight
+- Has parallels to ideas outside of Rust, building on external knowledge that might exist
+- Easy for cargo to parse and modify
+  - Cargo needs a simple syntax subset to parse to ensure it can adequately report errors for the part it cares about but does not get in the way of rustc reporting the high quality errors it cares about
+- In the future, this can be leveraged by other build systems or tools
+
+### Alternative 1: Doc-comment
+
+```rust
+#!/usr/bin/env cargo
+
+//! ```cargo
+//! [package]
+//! edition = "2018"
+//! ```
+
+fn main() {
+}
+```
+
+Benefits
+- Parsers are available to make this work (e.g. `syn`, `pulldown-cmark`)
+- Familiar syntax both to read and write.
+  - When discussing with a Rust author, it was pointed out many times people preface code with a comment specifying the dependencies ([example](https://github.com/dtolnay/prettyplease#example)), this is the same idea but reusable by cargo
+  - When discussing on forums, people expressed how they had never seen the syntax but instantly were able to understand it
+
+Downsides:
+- When discussing with a Rust crash course teacher, it was felt their students would have a hard time learning to write these manifests from scratch
+  - Having the explain the overloading of concepts to new users
+  - Unpredictable location (both the doc comment and the cargo code block within it)
+  - Visual clutter (where clutter is overwhelming already in Rust)
+- Might be a bit complicated to do edits (translating between location within
+  `toml_edit` spans to the location within `syn` spans especially with different comment styles)
```

Comment:
do we necessarily have to use `syn` for this? if we only allow `//! ` then we could "just" filter out lines beginning with that prefix and in the `toml_edit` spans do `col += 4`

Reply:
- If we only allow `//!` then they are no longer doc comments
  - When discussing doc comments, not everyone is agreed on a single style
- There can still be parse ambiguities, like doc comments that appear in string

Reply:
> * If we only allow `//!` then they are no longer doc comments

can you expand on that? I'm not sure I understand that point

> * There can still be parse ambiguities, like doc comments that appear in string

I guess that's true. seems unlikely, but not technically impossible :D

Reply:
> > If we only allow //! then they are no longer doc comments
>
> can you expand on that? I'm not sure I understand that point

If we deviate from the [syntax of doc comments](https://doc.rust-lang.org/reference/comments.html#doc-comments), even if its a subset, then it becomes a thing of its own.